### ISSUE TITLE
Instant View support

### ISSUE THREAD

**Issue by Marocco2** [telegramdesktop/tdesktop#3492]
Please add support to IV articles
![image](https://cloud.githubusercontent.com/assets/3944248/26797173/49e0cce0-4a2c-11e7-9f1f-5cd2c4b82e7d.png)

**Comment by MHTE** [comment]
Instant view was made to provide ease of access using built in browsers in each operating system. Telegram desktop is built to be multi platform so if they want to add it they may need to include some kind of browser which I don't think would be easy.
Safari and Firefox already provide a similar function.

**Comment by stek29** [comment]
@MHTE Instant view is not in HTML -- it's in tg's special format. However, IV pages can have "iframe" like parts -- for youtube, twitter, and a couple of other selected sites. 
So it is possible to render IV pages without a browser.

**Comment by elswerky** [comment]
yes please ,,,,why don't you ask durov on telegram to implement it ??

**Comment by stek29** [comment]
@elswerky because it'll be very annoying for Durov?
Problem with instant view are embeds, which would require TDesktop to either have a bundled browser (hi issues, vulns, even more complicated build process and larger binary size), or to handle API of every service which can be embedded (even worse)

**Comment by simple-elf** [comment]
Instant view works well on telegram desktop for MacOS. But doesn't work in windows desktop client. Screenshot from first post shows, how it looks on windows client.

**Comment by stek29** [comment]
@simple-elf no, it doesn't work on Telegram **Desktop** for macOS, only on Telegram native macOS app which is different from Telegram Desktop.

**Comment by AntonBoch1244** [comment]
@simple-elf That screenshot show look in Telegram Desktop application on every OS where Telegram Desktop client is available.

IV can be available on Telegram Desktop but for very high cost of performance(if using show in in-app web browser) on low-end machines(e.g. Desktop with motherboard Foxconn MXPG31 and CPU Intel Celeron E3400 on Socket 775).
i.e. how IV works step-by-step:
1. Parse HTML code
2. Get text information and images
3. Compile information to humanreadable form
4. Show information to user
That method cost more RAM but this faster method if site not blocked some information of course.
Or we can just open web browser and get same information without rewriting.

Sorry for my english and some errata in explanations.

**Comment by stek29** [comment]
@AntonBoch1244 No, that's not how IV works. You're very very wrong.

IV is generated on server side.

But embeds will be painful. That's the only problem -- embeds in IV are normal web pages, which have to be parsed and displayed by client.

**Comment by AntonBoch1244** [comment]
@stek29, well me need wash my mind about that then.
About IV please explain how it works if am i not right.

**Comment by AntonBoch1244** [comment]
@stek29 i check how IV works.
And drop my comments, as well. Because fully wrong answer in my now removed comments.

**Comment by Aokromes** [comment]
Maybe this can help to implement IV:
http://doc.qt.io/qt-5/qtwebview-minibrowser-example.html

**Comment by kirsan31** [comment]
Qt WebView can't be statically build, so no way for now...

**Comment by Piterden** [comment]
iv of this issue
![image](https://user-images.githubusercontent.com/5930429/40203577-5022d784-5a2e-11e8-8d5d-9119ab4a6b12.png)

**Comment by mugavri** [comment]
> Qt WebView can't be statically build, so no way for now...

So I think Telegram should implement an instant view site that gets the link to a site that processes it on a server side, then returns it for viewing from any browser.


https://t.me/iv?url=https%3A%2F%2Ftelegra.ph%2Fapi&rhash=3aae0ab195c808

and this link should by used by Qt WebView

**Comment by alxlg** [comment]
@stek29 since many pages don't have iframes, why not just render the rest and in case there is an iframe replace it with a warning that the content can't be diplayed and a button to open the page in a browser?

**Comment by glenfordwilliams** [comment]
any updates on this? telegram desktop really needs instant view

**Comment by Aokromes** [comment]
maybe now that telegram desktop uses QT 5.12 WebEngine can be used to implement this.

**Comment by kirsan31** [comment]
> maybe now that telegram desktop uses QT 5.12 WebEngine can be used to implement this.

I am afraid that QtWebEngine still can't be build statically... :(

**Comment by alxlg** [comment]
@kirsan31 could you please tell me why my suggestion is not an option? This one:

> since many pages don't have iframes, why not just render the rest and in case there is an iframe replace it with a warning that the content can't be diplayed and a button to open the page in a browser?

**Comment by kirsan31** [comment]
@alxlg 
The problem not only in iframes, the problem is that we have no html render control (browser) in Qt, that can be used in TD.
`QtWebEngine` can't be build statically and other Qt controls have [very limited](https://doc.qt.io/qt-5/richtext-html-subset.html) html support...

**Comment by alxlg** [comment]
@kirsan31 why do you need HTML support? Can't the text and the images, cached on the server as Instant View, be rendered without involving HTML? Does Telegram for Android use a web engine for Instant View? It doesn't look so...

**Comment by AndydeCleyre** [comment]
I'm not sure where I can examine IV documents directly, but maybe it would make sense to fetch and then convert to markdown, then render with `QTextDocument` or `QTextBrowser`. While they support some html, maybe limiting to supported element types would be simpler this way :man_shrugging:.

From [New Features in Qt 5.14](https://wiki.qt.io/New_Features_in_Qt_5.14):

>- QTextDocument/QTextTable now support

>     - per-edge border styling via QTextTableCellFormat
>     - border-collapse mode
>     - HTML table style import (partial) and export

>- QTextDocument supports reading and writing Markdown format, as an alternative to HTML. If you read HTML and write Markdown, or vice-versa, the formatting should be preserved to the extent that the CommonMark and GitHub specs allow (including headings, tables, bullet lists, block quotes and code blocks); but we don't guarantee all cases yet, because it's thinly tested so far.

>- Qt Quick

>     - Text and TextEdit now support Markdown format (CommonMark and GitHub dialects) as an alternative to HTML.
>        - Includes the GitHub checklist extension, such that you can click to toggle checkboxes in a TextEdit.

**Comment by glenfordwilliams** [comment]
Still looking forward to giving property iv support in desktop

**Comment by AhyaH-YHWH** [comment]
So a mini web browser can't be added for all TDesktop Clients?

**Comment by mussolene** [comment]
IV in TG Desktop is a pretty good feature. if you add this, many security issues will be closed. Because the opening of the site can be controlled in the application

**Comment by mussolene** [comment]
maybe add a python browser to the app or the chromium kernel

**Comment by Cajuteq** [comment]
I agree it would be very nice to have instant view in desktop and think that the markdown option is quite good as I (we?) usually use instant view for article reading and are almost only interested in text (maybe images when those are charts)

**Comment by Aokromes** [comment]
better later than never https://github.com/telegramdesktop/tdesktop/commit/a888b8d846cab7d53b63dd3e31f388ba340e9806

**Comment by alxlg** [comment]
Is it planned for Linux too?

**Comment by AndydeCleyre** [comment]
Yeah @Aokromes  can we keep this open until the feature reaches all supported platforms, rather than having someone create a new issue with the same request?

**Comment by john-preston** [comment]
@alxlg I'm not sure. It is done using WebView and it's very unreliable on Linux and requires big amount of coding to support Instant View on top of that. So right now it's not a priority for me.

**Comment by john-preston** [comment]
(on Linux)

**Comment by ilya-fedin** [comment]
Linux instant view implementation was progressing quite slowly, mainly due to media content weirdly not loading. So far it was assumed that the attempts to implement [range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Range_requests) are wrong due to no developer with a good webkitgtk knowledge on the project. After more than a year of research, it seems to be actually [webkitgtk not supporting media with custom URI schemas](https://bugs.webkit.org/show_bug.cgi?id=146351) [nor supporting intercepting HTTPS requests](https://webkit-gtk.webkit.narkive.com/mfmjYtC6/web-context-overriding-handling-of-http-s-schemes) which means implementing instant view on Linux is impossible (at least with @john-preston's requirement to support it with media or not at all). FWIW, this seem to [also affect Tauri](https://github.com/tauri-apps/tauri/issues/3725).
